### PR TITLE
Support different reset modes during import

### LINKED ISSUE #23: Support running from an empty database without dropping/re-creating
Devdata is a very useful tool, however in some cases we don't want to drop the underlying database. This happens for example when using devdata to manage a hosted staging environment, where the creation of the database itself is (deliberately) outside of the permissions set available to the user being used to run migrations & devdata.

It would be great if devdata supported this.

Maybe an flag to `devdata_import` which instead of dropping the database checks that there are no tables present (and complains if there are)?

### PR DESCRIPTION
**Pull request by PeterJCLaw**
This introduces support for "reset modes"[^1] -- different approaches for ensuring a clean database before importing the fresh data. Initial support is for three modes:

- `drop-database`: the default; drops the database & re-creates it.
- `drop-tables`: drops the tables the Django codebase is aware of, useful if the Django database user doesn't have access to drop the entire database.
- `none`: no attempt to reset the database, useful if the user has already manually configured the database or otherwise wants more control over setup.

Using the second of these in a staging-type environment is the main motivation for this change.

This PR also updates the existing `PostgresSequences` extra to cope with a sequence of a given name already being present by overwriting it. This doesn't feel like the best solution (the result if the import somehow lists the same sequence twice may be unexpected), but seemed probably the simplest for now given what devdata is typically used for.

Review by commit may be useful, they're in a hopefully sensible order for that.

Testing is achieved by adding reset modes to all existing import tests as well as creating a few more. I've also tested this against the use-case I originally had (importing against an existing database part of a staging environment).

Fixes https://github.com/danpalmer/django-devdata/issues/23

TODO:
- [x] :lady_beetle: ensure the django migrations table is cleared when using the `drop-tables` mode -> tested & fixed in 42271396d13b9448c8545126dcf45f6dc894e1b5
- [x] validate behaviour of `none` when importing over the top of an existing matching schema -> test demonstrating result in 84a2b03b6e6dfe857f360a093664bee49566abf6

[^1]: This doesn't feel like the best name, however "strategies" is already taken in this project and I didn't want to confuse that. Open to changing this if we can come up with something better!

### CODE REVIEW

**Comment by PeterJCLaw** [comment]
@lirsacc this may be of interest as I think you've wanted this before too.

File: src/devdata/reset_modes.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        connection = connections[django_dbname]
+
+        with connection.cursor() as cursor:
+            table_infos = connection.introspection.get_table_list(cursor)
+            models = connection.introspection.installed_models(
+                [x.name for x in table_infos],
+            )
+            with connection.schema_editor() as editor:
+                for model in models:
+                    editor.delete_model(model)
```
Comment (danpalmer) [inline src/devdata/reset_modes.py]: Does this run into deletion ordering issues?

**Comment by danpalmer** [comment]
Thanks for this Peter! I've not done a line by line review yet, I need to let this one stew for a few days if that's ok? The problem makes sense, having a solution in this codebase also makes sense, and this looks like a reasonable option.

A few questions/alternatives to discuss....

- The use-case of the staging environment makes sense, but I didn't feel the use case was clear from the description of how to use it in the readme. The default is nicely labelled, but I wonder if we can come up with a more compelling decision tree or something where the motivation for all three options is obvious, such that there's no confusion on which is the best option.
- Related, what are the downsides of using the per-table approach? Unless ridiculous, performance isn't a concern here I don't think. Do we need a full DB drop option?
-  What would this look like if it was a property of the strategy – i.e. whether the strategy wants its underlying data reset or not? Strategies may be idempotent or not (I think?). Would it be simpler if they always were or weren't?
- Can we unify the reset modes and the sequences behaviour? Alternatively, do they need their own reset modes – drop/rewind?
- Should this be a command line flag or a setting, or both?

File: src/devdata/reset_modes.py (hunk 1 above)
Comment (PeterJCLaw) [reply src/devdata/reset_modes.py]: I'm hoping this is safe with regards to e.g: foreign keys -- that was part of the motivation to use the schema editor. When I had a quick look I thought it should handle references for us, though I probably need to do a deeper pass through its code and/or add some more tests to be really confident. I perhaps ought to also run this against a large codebase which might show up some edge cases :)

**Comment by PeterJCLaw** [comment]
> Thanks for this Peter! I've not done a line by line review yet, I need to let this one stew for a few days if that's ok?

Yup, sure.

> The problem makes sense, having a solution in this codebase also makes sense, and this looks like a reasonable option.
> 
> A few questions/alternatives to discuss....

Good thoughts :+1:, thanks for the prompt review :)

> * The use-case of the staging environment makes sense, but I didn't feel the use case was clear from the description of how to use it in the readme. The default is nicely labelled, but I wonder if we can come up with a more compelling decision tree or something where the motivation for all three options is obvious, such that there's no confusion on which is the best option.

:+1: will have a think.

> * Related, what are the downsides of using the per-table approach? Unless ridiculous, performance isn't a concern here I don't think.

I think the main one is that it assumes that the existing database is close enough to what the current codebase defines that it can use the current models to empty the database. There are various ways that might not be true (even just being on a different branch might hit it) which could either fail loudly or not clear things out that the user might expect would be removed.
Sequences are the main example which come to mind for the non-failing case, though other tables also feel possible. I don't know what else might be left behind (especially on databases other than Postgres).

One thing which has just occurred to me and I suspect isn't covered by tests here (and is almost certainly a :lady_beetle:) -- the django migrations table isn't being cleared out, so that'll need addressing.

> Do we need a full DB drop option?

Aside from the limitations of the alternatives, I think there's a couple of reasons to keep the `drop-database` option:
- it makes this a non-breaking change (at least when kept as the default), and
- it's potentially handy in cases there may not be an existing database (e.g: getting started in local dev).

Note that neither the `drop-tables` nor the `none` options handle the lack of an existing database currently. From what I recall you mentioned at one point that it had been challenging to detect an existing database (I think this was in the context of why there's confirmation around dropping the database even if there isn't one), so I didn't look into handling that here. It may be that this has gotten easier now though, so perhaps there's more which could be done.

> * What would this look like if it was a property of the strategy – i.e. whether the strategy wants its underlying data reset or not? Strategies may be idempotent or not (I think?). Would it be simpler if they always were or weren't?

Interesting, I hadn't thought about pushing this down to that layer.

This might fit alongside the `none` or `drop-tables` approaches (or variations thereof), though obviously not with the `drop-database` mode. To me this feels like it would end up pushing towards removing `drop-database` and maybe encouraging more complexity in the strategies to make them idempotent.

Thinking it through, another source of complexity would be handling the case of a single model with several strategies. For me this feels like it'd be getting too complex and having the reset separate is going to be quite a bit simpler. I guess it depends on how advanced you want this tool to get in terms of what it offers around working with an existing database.

> * Can we unify the reset modes and the sequences behaviour? Alternatively, do they need their own reset modes – drop/rewind?

Perhaps. Feels like this might need a bit more thinking -- currently there's no relationship between these modes and the later processing. I'll admit sequences weren't something I'd thought about until actually building this, so they ended up getting a simple tweak rather than being designed for explicitly.

> * Should this be a command line flag or a setting, or both?

Hadn't thought about making it a setting. While my use-case is one where I mostly want a different mode in different environments, it's not clear to me that that would always be the case nor that that's necessarily a good motivation for a setting. My immediate thought is that this feels like a thing you want to control when you run the command (or write the wrapper script) rather than something that's a property of the codebase.
Things like django-configurations do allow settings to be properties of the environment but I think I'd (personally) still find I wanted the CLI flag.

File: README.md [hunk 2]
```diff
@@ -124,12 +124,29 @@ step.
...
 exporting strategies have been used those must have run first, or their outputs
 must have been downloaded if they are being shared/hosted somewhere.
 
 Factory-based strategies generate data during this process.
 
+##### Reset modes
+
+By default any existing database will be removed, ensuring that a fresh database
+is created for the imported data. This is expected to be the most common case
+for local development, but may not always be suitable.
```
Comment (danpalmer) [inline README.md:149]: Can we add an example of the command line argument being used here, just to demonstrate that this is controlled in that way? Doesn't have to be right here, just in an appropriate place around here.

File: src/devdata/extras.py [hunk 3]
```diff
@@ -155,6 +155,19 @@ def check_simple_value(mapping: Dict[str, str], *, key: str) -> str:
...
+                # to remove sequences too, however that's likely complicated and
+                # it's easy enough to handle here.
+                #
+                # Sequences don't nicely fit into one of just schema or data,
+                # they're somewhat inherently both. Given that Django's
+                # "loaddata" over-writes existing rows in tables, it seems
+                # reasonable to do something similar for sequences -- even if
+                # that means we actually drop the sequence and fully re-create
+                # it.
+                cursor.execute(f"DROP SEQUENCE IF EXISTS {name}")
```
Comment (danpalmer) [inline src/devdata/extras.py:181]: Should we only do this for reset modes that don't drop the database? Should we add a hook in the reset mode for resetting the sequence?

File: src/devdata/management/commands/devdata_import.py [hunk 4]
```diff
@@ -30,31 +31,47 @@ def add_arguments(self, parser):
...
-                "Are you sure you want to continue? [y/N]: ".format(
-                    self.style.WARNING(database),
-                    self.style.WARNING(settings.DATABASES[database]["NAME"]),
-                    self.style.WARNING(socket.gethostname()),
-                ),
-            ).lower()
-            != "y"
+        if (
+            not no_input
+            and reset_mode.requires_confirmation
```
Comment (danpalmer) [inline src/devdata/management/commands/devdata_import.py]: What do you think about pushing this down into the reset mode? I feel like we should prompt the user in all cases – dropping the database, dropping the tables, and blatting a bunch of new data into an existing database could all be bad.

**Review by danpalmer — APPROVED** [review]
Re: pushing down into strategies...

> Thinking it through, another source of complexity would be handling the case of a single model with several strategies. For me this feels like it'd be getting too complex and having the reset separate is going to be quite a bit simpler.

Very good point, I think this answers this one entirely for me, thanks!

Re: command line flag...

> While my use-case is one where I mostly want a different mode in different environments, it's not clear to me that that would always be the case nor that that's necessarily a good motivation for a setting. My immediate thought is that this feels like a thing you want to control when you run the command (or write the wrapper script) rather than something that's a property of the codebase.

Sounds good. Let's start with a command-line flag. I think you're right that this is a property of the particular development environment and the current aim of the developer at a point in time, rather than a configuration around the runtime environment that might be configured with something like Configurations. We can always add a setting later if there's any desire for it.

- - -

Thanks for all your time on this and on replying in detail to the comments! Let's go with the approach in this PR. I've done a line by line review now.

Don't forget the potential bug about the handling of the migrations table, if it is necessary.

File: src/devdata/management/commands/devdata_import.py (hunk 4 above)
Comment (PeterJCLaw) [reply src/devdata/management/commands/devdata_import.py]: If we're going to always prompt (which seems reasonable), that feels to me like something which would be better served by being outside the modes (i.e: as currently) to avoid duplication.
For context: I introduced `requires_confirmation` purely for the `none` mode and don't expect any other modes to need to change it. I agree that even then doing the import could potentially be destructive.

As a concrete suggestion: how about removing `requires_confirmation` and having the `none` mode have a description something like "import into of the current state of the database" rather than bypassing the check?
(I'll try to come up with wordings which make sense for each of the modes & possible database states [without detection of the the latter], then work out how much of the description needs to live in the modes).

(Another small reason I didn't push the interaction down into the mode classes is that if there were to ever be a programmatic interface to devdata, then having `input` calls in the modes would need reworking; having that in the command feels less likely to be surprising)

File: src/devdata/extras.py (hunk 3 above)
Comment (PeterJCLaw) [reply src/devdata/extras.py:181]: Maaybe. This is probably the part of this PR I'm least happy with. I considered hooking into the reset modes here, however it's not clear how that would work especially given that this extra is Postgres specific while the reset modes are (I believe mostly) sufficiently general to any of Django's supported backends[^1].

To clarify: is your concern here with the possible outcomes under some combination(s) of existing state & reset mode, or with the approach to solving the problem? (or both?)

Running through my thinking on the interactions with each mode:
- `drop-database`: nothing needed (sequence already gone)
- `drop-tables`: should probably drop the sequence (as this code does)
- `none`: unclear; leaving an existing sequence in place would always error during import, while an import over an empty database would just work. In contrast, I'm expecting (though haven't tested and should) that `none` against an existing matching database with the right schema (and migrations table) would import cleanly (modulo conflicts with the existing data).

I _think_ that the behaviour this code achieves is approximately the same as equivalent schema/data imports and probably most likely what the user wants (regardless of mode). Agree it doesn't feel like the cleanest way to achieve that though!

[^1]: I suspect that some of the `IF EXISTS` usage which is now part of the reset modes but was already present may in fact be Postgres specific, though this was an existing limitation.

File: README.md (hunk 2 above)
Comment (PeterJCLaw) [reply README.md:149]: Thoughts on ~beef078~ 219e81018350f9b3fe3293746ad26940434dd360?

File: src/devdata/management/commands/devdata_import.py (hunk 4 above)
Comment (PeterJCLaw) [reply src/devdata/management/commands/devdata_import.py]: Implemented that suggestion in 3ad2e85476fb3344042d86d87e20a78f9055fbfd.

File: src/devdata/extras.py (hunk 3 above)
Comment (PeterJCLaw) [reply src/devdata/extras.py:181]: I've added some docs around this as part of 729a19f3a0430df73e75b022f77a4373746da226 (which also adds more docs for the reset modes).

**Comment by PeterJCLaw** [comment]
@danpalmer thanks for the review. I think I've addressed everything now -- both the TODOs and the review comments, so would be great if you had time for another pass. Hopefully review by commit of the new changes is clear enough, but happy to answer any further questions you have.

**Comment by PeterJCLaw** [comment]
Ah, just for clarity: I still haven't done any manual testing here and probably won't have time this side of the new year. The updates since the first review do include better testing though so I'm a little more confident in how this will behave now. Happy to wait & do some manual testing though if that's desired.

**Comment by PeterJCLaw** [comment]
@danpalmer hope you've had a good break, do you know when you might have a chance to have another look at this PR?

File: src/devdata/extras.py (hunk 3 above)
Comment (danpalmer) [reply src/devdata/extras.py:181]: Makes sense, the docs sound good! Thanks!

File: src/devdata/management/commands/devdata_import.py (hunk 4 above)
Comment (danpalmer) [reply src/devdata/management/commands/devdata_import.py]: Perfect!

**Review by danpalmer — APPROVED** [review]
Thanks so much for this @PeterJCLaw. Go ahead and merge when you're ready!

File: src/devdata/reset_modes.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            table_names = connection.introspection.table_names(cursor)
+
+            models = connection.introspection.installed_models(table_names)
+
+            if MigrationRecorder(connection).has_table():
+                models.add(MigrationRecorder.Migration)
+
+            with connection.schema_editor() as editor:
+                for model in models:
+                    editor.delete_model(model)
```
Comment (PeterJCLaw-mns) [inline src/devdata/reset_modes.py:106]: Hrm, should this be transactional? That way we'd either drop all the tables or none, which is probably better than maybe dropping some?

A failure mode here is if the user has permission for some of the tables but not all.

Given that we're not dropping the database it might even be nice to have the whole import be transactional, but that likely requires deeper changes.

Comment (PeterJCLaw) [inline src/devdata/reset_modes.py:106]: Hrm, should this be transactional? That way we'd either drop all the tables or none, which is probably better than maybe dropping some?

A failure mode here is if the user has permission for some of the tables but not all.

Given that we're not dropping the database it might even be nice to have the whole import be transactional, but that likely requires deeper changes.

Comment (PeterJCLaw) [inline src/devdata/reset_modes.py:106]: A progress bar here might be nice. And/or some generic feedback that the "reset" portion of the import has completed.

**Comment by PeterJCLaw** [comment]
Have now tested this in the use-case I originally had in mind, it seems to work :)

File: src/devdata/reset_modes.py (hunk 5 above)
Comment (PeterJCLaw) [reply src/devdata/reset_modes.py:106]: I'm going to leave this as-is for the moment -- following the pattern of the drop-database mode which can also leave things part complete (even though that's perhaps less likely).

Comment (PeterJCLaw) [reply src/devdata/reset_modes.py:106]: I'm going to leave this as-is. I still think that a progress report here would be good, however making it clear that this is on the reset side feels non-trivial without some more general status reporting during import.
